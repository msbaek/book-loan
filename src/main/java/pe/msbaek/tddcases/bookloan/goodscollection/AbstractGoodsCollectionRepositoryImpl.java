package pe.msbaek.tddcases.bookloan.goodscollection;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

@RequiredArgsConstructor
public abstract class AbstractGoodsCollectionRepositoryImpl implements GoodsCollectionRepository {
    private final GoodsCollectionJpaRepository jpaRepository;

    @Override
    public List<GoodsCollection> findByNameContaining(String name, Pageable pageable) {
        return jpaRepository.findByNameContaining(name, pageable);
    }

    @Override
    public Optional<GoodsCollection> findById(Long id) {
        return jpaRepository.findById(id);
    }

    @Override
    public GoodsCollection save(GoodsCollection goodsCollection) {
        return jpaRepository.save(goodsCollection);
    }
}
