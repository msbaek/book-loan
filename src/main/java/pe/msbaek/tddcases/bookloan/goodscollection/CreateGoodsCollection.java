package pe.msbaek.tddcases.bookloan.goodscollection;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.jmolecules.architecture.hexagonal.PrimaryAdapter;
import org.springframework.graphql.data.method.annotation.Argument;
import org.springframework.graphql.data.method.annotation.MutationMapping;
import org.springframework.stereotype.Controller;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;

import java.util.List;

@Slf4j
@RequiredArgsConstructor
@Transactional
@Controller
@PrimaryAdapter
public class CreateGoodsCollection {
    @MutationMapping("createGoodsCollection")
    public Long createGoodsCollection(@Argument final CreateGoodsCollectionRequest request) {
        log.info("request: {}", request);
        return 1l;
    }

    record CreateGoodsCollectionRequest(String name, List<String> ids) {
    }
}