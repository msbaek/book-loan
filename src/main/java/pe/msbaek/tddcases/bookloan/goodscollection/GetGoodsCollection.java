package pe.msbaek.tddcases.bookloan.goodscollection;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.jmolecules.architecture.hexagonal.PrimaryAdapter;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.graphql.data.method.annotation.Argument;
import org.springframework.graphql.data.method.annotation.QueryMapping;
import org.springframework.stereotype.Controller;
import org.springframework.transaction.annotation.Transactional;
import pe.msbaek.tddcases.bookloan.common.SearchDto;

import java.util.List;

@Slf4j
@RequiredArgsConstructor
@Transactional(readOnly = true)
@Controller
@PrimaryAdapter
public class GetGoodsCollection {
    @QueryMapping("pagedGoodsCollection")
    public Page<GoodsCollectionDto> pagedGoodsCollection(@Argument final SearchDto request) {
        List<GoodsCollectionDto> result = List.of(
                new GoodsCollectionDto(1L, "상품종", 1L, "2022-01-01 00:00:00", 1L, "2022-01-01 00:00:00")
                , new GoodsCollectionDto(2L, "상품종2", 1L, "2022-01-01 00:00:00", 1L, "2022-01-01 00:00:00")
                , new GoodsCollectionDto(3L, "상품종3", 1L, "2022-01-01 00:00:00", 1L, "2022-01-01 00:00:00")
                , new GoodsCollectionDto(4L, "상품종4", 1L, "2022-01-01 00:00:00", 1L, "2022-01-01 00:00:00")
                , new GoodsCollectionDto(5L, "상품종5", 1L, "2022-01-01 00:00:00", 1L, "2022-01-01 00:00:00")
        );
        long total = result.size();
        return new PageImpl<>(result, PageRequest.of(request.page(), request.size()), total);
    }

    public record GoodsCollectionDto(Long id, String name, Long createdBy, String createdAt, Long updatedBy, String updatedAt) {
    }
}