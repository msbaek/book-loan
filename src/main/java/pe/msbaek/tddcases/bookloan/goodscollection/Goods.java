package pe.msbaek.tddcases.bookloan.goodscollection;

public record Goods(Long goodsNo, String goodsId, String barcode) {
}
