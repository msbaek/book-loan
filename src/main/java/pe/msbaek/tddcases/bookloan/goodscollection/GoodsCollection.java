package pe.msbaek.tddcases.bookloan.goodscollection;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Comment;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@NoArgsConstructor
@Getter
@Entity
@Table(name = "GOODS_COLLECTION")
@Comment("특정 상품군")
public class GoodsCollection {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Comment("특정 상품군 ID")
    private Long id;
    @Column(name = "NM")
    @Comment("특정 상품군 이름")
    private String name;
    @Column(name = "CREATED_BY")
    @Comment("특정 상품군 생성한 사용자 ID")
    private Long createdBy;
    @Column(name = "CREATED_AT")
    @Comment("특정 상품군 생성일")
    private LocalDateTime createdAt;
    @Column(name = "UPDATED_BY")
    @Comment("특정 상품군 수정한 사용자 ID")
    private Long updatedBy;
    @Column(name = "UPDATED_AT")
    @Comment("특정 상품군 수정일")
    private LocalDateTime updatedAt;
    @OneToMany(mappedBy = "goodsCollection", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Comment("특정 상품군에 속한 상품들")
    private List<GoodsCollectionItem> goodsCollectionItems = new ArrayList<>();

    public GoodsCollection(long id, String name, long createdBy, LocalDateTime createdAt) {
        this.id = id;
        this.name = name;
        this.createdBy = createdBy;
        this.createdAt = createdAt;
    }

    public GoodsCollection(String name, Long createdBy) {
        this.name = name;
        this.createdAt = LocalDateTime.now();
        this.createdBy = createdBy;
    }

    public void add(GoodsCollectionItem item) {
        goodsCollectionItems.add(item);
    }

    public void setId(Long id) {
        this.id = id;
    }
}
