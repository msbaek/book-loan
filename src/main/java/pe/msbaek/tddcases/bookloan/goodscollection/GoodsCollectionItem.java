package pe.msbaek.tddcases.bookloan.goodscollection;

import lombok.Getter;

@Getter
public class GoodsCollectionItem {
    private Long id;
    private Long goodsNo;
    private String goodsId;
    private String barcode;

    public static GoodsCollectionItem of(Goods goods) {
        GoodsCollectionItem item = new GoodsCollectionItem();
        item.goodsNo = goods.goodsNo();
        item.goodsId = goods.goodsId();
        item.barcode = goods.barcode();
        return item;
    }

    public void setId(Long id) {
        this.id = id;
    }
}