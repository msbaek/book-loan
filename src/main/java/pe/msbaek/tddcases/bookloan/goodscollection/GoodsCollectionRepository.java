package pe.msbaek.tddcases.bookloan.goodscollection;

import org.springframework.data.domain.Pageable;

import java.util.List;

public interface GoodsCollectionRepository {
    List<GoodsCollection> findByNameContaining(String name, Pageable pageable);
}
