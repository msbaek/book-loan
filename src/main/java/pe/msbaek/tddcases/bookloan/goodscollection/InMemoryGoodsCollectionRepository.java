package pe.msbaek.tddcases.bookloan.goodscollection;

import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.toMap;

public class InMemoryGoodsCollectionRepository implements GoodsCollectionRepository {
    static final Map<Long, GoodsCollection> goodsCollectionMap;
    static AtomicLong goodsCollectionIdGenerator = new AtomicLong(1);

    static {
        goodsCollectionMap = IntStream.range(0, 40)
                .mapToLong(i -> nextId())
                .mapToObj(id -> new GoodsCollection(id, "name" + id, 1L, LocalDateTime.now()))
                .collect(toMap(GoodsCollection::getId, goodsCollection -> goodsCollection));
    }

    private static Long nextId() {
        return goodsCollectionIdGenerator.getAndIncrement();
    }

    @Override
    public List<GoodsCollection> findByNameContaining(String name, Pageable pageable) {
        return goodsCollectionMap.values().stream()
                .filter(goodsCollection -> goodsCollection.getName().contains(name))
                .skip(pageable.getOffset())
                .limit(pageable.getPageSize())
                .toList();
    }
}