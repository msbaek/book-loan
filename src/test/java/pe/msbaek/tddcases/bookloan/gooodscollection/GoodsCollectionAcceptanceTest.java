package pe.msbaek.tddcases.bookloan.gooodscollection;

import com.ktown4u.utils.YamlPrinter;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.graphql.tester.AutoConfigureHttpGraphQlTester;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.graphql.test.tester.GraphQlTester;
import org.springframework.graphql.test.tester.HttpGraphQlTester;
import org.springframework.test.context.jdbc.Sql;
import pe.msbaek.tddcases.bookloan.goodscollection.GetGoodsCollection;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@AutoConfigureHttpGraphQlTester
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
public class GoodsCollectionAcceptanceTest {
    @Autowired private HttpGraphQlTester graphQlTester;

    @Test
    public void query_pagedGoodsCollection() throws Exception {
        String queryString = """
                query {
                  pagedGoodsCollection(request: {
                    keyword: "상품군",
                    type: "type",
                    page: 0,
                    size: 10
                  }) {
                    content {
                      id
                      name
                      goodsCollectionItems {
                        goodsNo
                        goodsId
                        barcode
                      }
                    }
                  }
               }
                """;

        List<GetGoodsCollection.GoodsCollectionDto> result = request(queryString, "pagedGoodsCollection.content")
                .entityList(GetGoodsCollection.GoodsCollectionDto.class)
                .get();
        Approvals.verify(YamlPrinter.printWithExclusions(result, "createdBy" ,"createdAt" ,"updatedBy" ,"updatedAt"));
    }

    @Test
    public void create_goods_collection() throws Exception {
        String queryString = """
                mutation {
                    createGoodsCollection(request: {
                        name: "상품군 이름"
                        ids: ["GD00111839", "GD00111838", "GD00111836"]
                    })
                }
                """;

        Long result = request(queryString, "createGoodsCollection")
                .entity(Long.class)
                .get();
        assertThat(result).isGreaterThan(0L);
    }

    private GraphQlTester.Path request(String queryString, String requestName) {
        return this.graphQlTester
                .mutate()
                .header("x-tester-no", "1001")
                .build()
                .document(queryString)
                .execute()
                .path(requestName);
    }
}