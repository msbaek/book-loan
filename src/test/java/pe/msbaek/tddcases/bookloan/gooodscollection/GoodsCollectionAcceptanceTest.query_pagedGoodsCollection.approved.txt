---
- id: 1
  name: "상품종"
- id: 2
  name: "상품종2"
- id: 3
  name: "상품종3"
- id: 4
  name: "상품종4"
- id: 5
  name: "상품종5"
